dnsprovider/inmemory.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dnsprovider
+
+import (
+	"errors"
+
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	"github.com/kubernetes-incubator/external-dns/plan"
+)
+
+var (
+	defaultType = "A"
+
+	// ErrZoneNotFound error returned when specified zone does not exists
+	ErrZoneNotFound = errors.New("specified zone not found")
+	// ErrRecordAlreadyExists when create request is sent but record already exists
+	ErrRecordAlreadyExists = errors.New("record already exists")
+	// ErrRecordNotFound when update/delete request is sent but record not found
+	ErrRecordNotFound = errors.New("record not found")
+	// ErrInvalidBatchRequest when record is repeated in create/update/delete
+	ErrInvalidBatchRequest = errors.New("record should only be specified in one list")
+)
+
+type zone map[string][]*InMemoryRecord
+
+// InMemory - dns provider only used for testing purposes
+// initialized as dns provider with no records
+type InMemory struct {
+	zones map[string]zone
+}
+
+// NewInMemory returns InMemory DNS provider interface
+func NewInMemory() *InMemory {
+	return &InMemory{
+		zones: map[string]zone{},
+	}
+}
+
+// InMemoryRecord - record stored in memory
+// has additional fields:
+// Type - type of string (TODO: Type should probably be part of endpoint struct)
+// Payload - string - additional information stored
+type InMemoryRecord struct {
+	Type    string
+	Payload string
+	endpoint.Endpoint
+}
+
+// Records returns the list of endpoints
+func (im *InMemory) Records(zone string) ([]endpoint.Endpoint, error) {
```

Comment:
I just used `*p` for provider in the other providers, so that I can also refer to it with `p`, regardless which exact provider it is. But that's obviously opinionated. Just want to keep the diff between different providers to a minimum.